reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,142 @@
 
 package kotlinx.coroutines.rx2
 
-import io.reactivex.Scheduler
+import io.reactivex.*
+import io.reactivex.disposables.*
+import io.reactivex.plugins.*
 import kotlinx.coroutines.*
-import java.util.concurrent.TimeUnit
-import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.channels.*
+import java.util.concurrent.*
+import kotlin.coroutines.*
 
 /**
  * Converts an instance of [Scheduler] to an implementation of [CoroutineDispatcher]
  * and provides native support of [delay] and [withTimeout].
  */
-public fun Scheduler.asCoroutineDispatcher(): SchedulerCoroutineDispatcher = SchedulerCoroutineDispatcher(this)
+public fun Scheduler.asCoroutineDispatcher(): CoroutineDispatcher =
+    if (this is DispatcherScheduler) {
+        dispatcher
+    } else {
+        SchedulerCoroutineDispatcher(this)
+    }
+
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Scheduler].
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    private val schedulerJob = SupervisorJob()
+
+    private val scope = CoroutineScope(schedulerJob + dispatcher + CoroutineExceptionHandler { _, throwable ->
+        RxJavaPlugins.onError(throwable)
+    })
+
+    override fun scheduleDirect(block: Runnable): Disposable =
+        scheduleDirect(block, 0, TimeUnit.MILLISECONDS)
+
+    override fun scheduleDirect(block: Runnable, delay: Long, unit: TimeUnit): Disposable {
+        if (!scope.isActive) return Disposables.disposed()
+        val newBlock = RxJavaPlugins.onSchedule(block)
+        return scope.launch {
+            if (delay > 0) {
+                delay(unit.toMillis(delay))
+            }
+            newBlock.run()
+        }.asDisposable()
+    }
+
+    override fun createWorker(): Worker =
+        DispatcherWorker(dispatcher, schedulerJob)
+
+    override fun shutdown() {
+        scope.cancel()
+    }
+
+    private class DispatcherWorker(dispatcher: CoroutineDispatcher, parentJob: Job) : Worker() {
+
+        private val workerJob = SupervisorJob(parentJob)
+        private val workerScope = CoroutineScope(workerJob + dispatcher)
+        private val blockChannel = Channel<SchedulerChannelTask>(Channel.UNLIMITED)
+
+        init {
+            workerScope.launch {
+                while (isActive) {
+                    val task = blockChannel.receive()
+                    task.execute()
+                }
+            }
+        }
```

Comment:
Running a (potentially) concurrent Job from the constructor is a bad style. It would be better to move the code from the `init {}` block in a separate `fun start()` and call `start()` from the `createWorker` after construction.